src/include/sof/audio/audio_stream.h

```diff
@@ -192,6 +228,82 @@ static inline void *audio_stream_wrap(const struct audio_stream *buffer,
 	return ptr;
 }
 
+/**
+ * Calculates available data in bytes, handling underrun_permitted behaviour
+ * @param stream Stream pointer
+ * @return amount of data available for processing in bytes
+ */
+static inline uint32_t
+audio_stream_get_avail_bytes(const struct audio_stream *stream)
+{
+	if (!stream->underrun_permitted)
+		return stream->avail;
+
+	return stream->avail != 0 ? stream->avail : stream->size;
+}
+
+/**
+ * Calculates available data in samples, handling underrun_permitted behaviour
+ * @param stream Stream pointer
+ * @return amount of data available for processing in samples
+ */
+static inline uint32_t
+audio_stream_get_avail_samples(const struct audio_stream *stream)
+{
+	return audio_stream_get_avail_bytes(stream) /
+		audio_stream_sample_bytes(stream);
+}
+
+/**
+ * Calculates available data in frames, handling underrun_permitted behaviour
+ * @param stream Stream pointer
+ * @return amount of data available for processing in frames
+ */
+static inline uint32_t
+audio_stream_get_avail_frames(const struct audio_stream *stream)
+{
+	return audio_stream_get_avail_bytes(stream) /
+		audio_stream_frame_bytes(stream);
+}
+
+/**
+ * Calculates free space in bytes, handling overrun_permitted behaviour
+ * @param stream Stream pointer
+ * @return amount of space free in bytes
+ */
+static inline uint32_t
+audio_stream_get_free_bytes(const struct audio_stream *stream)
+{
+	if (!stream->overrun_permitted)
```

Comment:
I would also deal with the overrun permitted here (next to it's conditional logic) than as the default return case. Needs comment too

Reply:
done